crates/oxc_linter/src/rules/eslint/no_shadow_restricted_names.rs

```diff
@@ -1,71 +1,179 @@
-    /// var undefined = "foo";
+    /// function NaN(){}
+    ///
+    /// !function(Infinity){};
+    ///
+    /// var undefined = 5;
+    ///
+    /// try {} catch(eval){}
     /// ```
     NoShadowRestrictedNames,
     correctness
 );
 
-static RESTRICTED: [&str; 5] = ["undefined", "NaN", "Infinity", "arguments", "eval"];
-
-fn safely_shadows_undefined(symbol_id: SymbolId, ctx: &LintContext<'_>) -> bool {
-    let symbol_table = ctx.semantic().symbols();
-    if symbol_table.get_name(symbol_id).as_str() == "undefined" {
-        let mut no_assign = true;
-        for reference in symbol_table.get_resolved_references(symbol_id) {
-            if reference.is_write() {
-                no_assign = false;
+fn binding_pattern_is_global_obj(
+    pat: &BindingPattern,
+    ignore_undefined: bool,
+) -> Option<(Atom, Span)> {
+    match &pat.kind {
+        oxc_ast::ast::BindingPatternKind::BindingIdentifier(boxed_bind_identifier) => {
+            if ignore_undefined && boxed_bind_identifier.name.as_str() == "undefined" {
+            } else if PRE_DEFINE_VAR.contains_key(boxed_bind_identifier.name.as_str()) {
+                return Some((boxed_bind_identifier.name.clone(), boxed_bind_identifier.span));
+            }
+            None
+        }
+        oxc_ast::ast::BindingPatternKind::ObjectPattern(boxed_obj_pat) => {
+            let properties = &boxed_obj_pat.properties;
+            for property in properties {
+                if let Some(value) = binding_pattern_is_global_obj(&property.value, false) {
+                    return Some(value);
+                }
             }
+            boxed_obj_pat
+                .rest
+                .as_ref()
+                .and_then(|boxed_rest| binding_pattern_is_global_obj(&boxed_rest.argument, false))
         }
-        let mut no_init = false;
-        let decl = ctx.nodes().get_node(symbol_table.get_declaration(symbol_id));
-        if let AstKind::VariableDeclarator(var) = decl.kind() {
-            no_init = var.init.is_none();
+        oxc_ast::ast::BindingPatternKind::ArrayPattern(boxed_arr_pat) => {
+            for element in boxed_arr_pat.elements.iter() {
+                if let Some(value) =
+                    element.as_ref().and_then(|e| binding_pattern_is_global_obj(e, false))
+                {
+                    return Some(value);
+                }
+            }
+            boxed_arr_pat
+                .rest
+                .as_ref()
+                .and_then(|boxed_rest| binding_pattern_is_global_obj(&boxed_rest.argument, false))
+        }
+        oxc_ast::ast::BindingPatternKind::AssignmentPattern(boxed_assign_pat) => {
+            binding_pattern_is_global_obj(&boxed_assign_pat.left, false)
         }
-        return no_assign && no_init;
     }
-    false
+}
+
+fn get_nearest_undefined_declare_span(ctx: &LintContext) -> Option<Span> {
+    let nodes = ctx.nodes();
+    let mut span: Option<Span> = None;
+    for node in nodes.iter() {
```

Comment:
Oh, I just noticed that this is iterating all the AST nodes, it would be great if this can be improved, because this will hurt performance. @lovebaihezi

Reply:
Yes, I'm reading the code to find a way to per it, but not today, I need some time.

Reply:
Cool, please let me know if you need help.